src/garn-bin.js

```diff
@@ -1,86 +1,227 @@
+
+if (isInstalledGlobally) {
+  const localGarn = path.join(process.cwd(), 'node_modules', '.bin', 'garn' + execExt);
+  if (fs.existsSync(localGarn)) {
+    const childGarn = childProcess.spawn(localGarn, process.argv.slice(2), {
+      cwd: process.cwd(),
+      stdio: 'inherit',
+    });
+    childGarn.on('exit', exitCode => process.exit(exitCode));
+  } else {
+    console.log('Error! Garn does not seem to be installed in the current working directory.');
+    console.log('You are most likely executing Garn from the incorrect folder.');
+    console.log('Your working directory must contain a package.json file with Garn (@avensia-oss/garn) installed.');
+    process.exit(1);
+  }
+} else {
+  const buildsystemPathArgName = 'buildsystem-path';
+  if (!process.argv.find(v => v.includes(`--${buildsystemPathArgName}`))) {
+    const assumedBuildsystemPath = path.join(process.cwd(), 'buildsystem');
+    if (!fs.existsSync(assumedBuildsystemPath)) {
+      console.log('Error! No buildsystem folder exists at:', assumedBuildsystemPath);
+      console.log('You are most likely executing Garn from the incorrect folder.');
+      console.log('Your working directory should contain a folder called buildsystem that contains the Garn tasks.');
+      process.exit(1);
+    }
+    process.argv.push('--' + buildsystemPathArgName, assumedBuildsystemPath);
   }
-}
 
-const needsCompile = anyFileInManifestHasChanged();
-let writeMetaData = false;
+  const argv = minimist(process.argv.slice(2));
+
+  const buildsystemPath = argv[buildsystemPathArgName];
+  const buildCache = '.buildcache';
+  const buildCachePath = path.join(buildsystemPath, buildCache);
+  const buildCacheManifestPath = path.join(buildCachePath, '.manifest.json');
+  const basePath = path.join(buildsystemPath, '..');
+  let rootPath = basePath;
+  if (!fs.existsSync(path.join(basePath, 'packages', 'garn.cmd'))) {
+    const parentName = path.basename(path.join(basePath, '..'));
+    if (parentName === 'packages') {
+      rootPath = path.join(basePath, '..', '..');
+    }
+  }
 
-const skipCompile = 'asap' in argv && fs.existsSync(buildCacheManifestPath);
+  const shouldRestore = shouldRestoreNpmPackages(rootPath, buildCachePath);
+  if (shouldRestore === false) {
+    compileIfNeededAndRun(argv, rootPath, buildsystemPath, buildCache, buildCachePath, buildCacheManifestPath);
+  } else {
+    restoreNpmPackages(rootPath, shouldRestore);
+  }
+}
 
-if ((!skipCompile && needsCompile) || 'compile-buildsystem' in argv) {
-  writeMetaData = true;
-  console.log('Compiling buildsystem...');
-  compile();
+/**
+ * @param {string} rootPath
+ * @param {'yarn' | 'npm'} restorePackagesWith
+ */
+function restoreNpmPackages(rootPath, restorePackagesWith) {
+  /**
+   * @param {number} npmOrYarnExitCode
+   */
+  const restartGarn = npmOrYarnExitCode => {
+    if (npmOrYarnExitCode < 1) {
+      const garnPath = path.join(rootPath, 'node_modules', '.bin', 'garn' + execExt);
```

Comment:
It should probably copy over the changed `yarn.lock` or `package-lock.json` to the buildcache on a successful install. Right now it just installs endlessly :sweat_smile: